### ISSUE TITLE
Undefined Index Notice - PHP

### ISSUE THREAD

**Issue by tbillington** [google/protobuf#2743]
I'm getting PHP notices for a lot of my types, not sure why.

```
PHP Notice:  Undefined index: .foo.bar.baz in /Users/trent/src/github.com/sajari/phptest/vendor/google/protobuf/php/src/Google/Protobuf/Internal/DescriptorPool.php on line 125
```

**Comment by TeBoring** [comment]
which version are u using?

**Comment by tbillington** [comment]
php grpc library: `"grpc/grpc": "v1.1.2"`
protoc: `libprotoc 3.2.0`

Just got the grpc_php_plugin from master a few hours ago.

**Comment by TeBoring** [comment]
Have you used protoc to generate code?

**Comment by tbillington** [comment]
Yup

```
protoc -I$PROTO_SRC --php_out=$dest --grpc_out=$dest --plugin=protoc-gen-grpc=/Users/trent/grpc/bins/opt/grpc_php_plugin $dir/*.proto
```

**Comment by TeBoring** [comment]
What's the protoc version?
And did you autoload your generated proto files?

**Comment by tbillington** [comment]
Yes, I autoloaded. I posted just before your previous comment the protoc version (maybe harder to see in email): `libprotoc 3.2.0`

**Comment by TeBoring** [comment]
Can I see your proto file and test code?

**Comment by tbillington** [comment]
Seems this is only happening on enums.

**Comment by TeBoring** [comment]
Maybe it can be fixed in you have a non-empty namespace. Let me know if it works.

**Comment by tbillington** [comment]
Still had the issue in a non-empty namespace.

**Comment by tbillington** [comment]
Here is some test code that demonstrates the issue. I removed the namespace because it had no impact.

Proto File:
```proto
syntax = "proto3";

package simpletest;

service SimpleTest {
  rpc Test (TestRequest) returns (TestResponse);
}

message TestRequest {
  enum TestEnum {
    ZERO = 0;
    ONE = 1;
  }

  string name = 1;

  TestEnum test_enum = 2;
}

message TestResponse {
  string reply = 1;
}
```

Test Code:
```php
<?php
use Grpc\ChannelCredentials;
require  __DIR__ . "/vendor/autoload.php";
include_once  __DIR__ . "/proto/SimpleTest/SimpleTestClient.php";
include_once  __DIR__ . "/proto/SimpleTest/TestRequest.php";
include_once  __DIR__ . "/proto/SimpleTest/TestRequest_TestEnum.php";
include_once  __DIR__ . "/proto/SimpleTest/TestResponse.php";
include_once  __DIR__ . "/proto/GPBMetadata/Messages.php";
$client = new SimpleTest\SimpleTestClient("localhost:1111", ['credentials' => Grpc\ChannelCredentials::createInsecure()]);
$req = new SimpleTest\TestRequest();
$req->setName("Foo");
$req->setTestEnum(SimpleTest\TestRequest_TestEnum::ONE);
try {
    list($res, $status) = $client->Test($req)->wait();
    printf("%s\n", $res->getReply());
} catch (Exception $e) {
    printf("%s\n", $e);
}
```

Output:
```
PHP Notice:  Undefined index: .simpletest.TestRequest.TestEnum in /Users/trent/src/github.com/sajari/proto-php-example/vendor/google/protobuf/php/src/Google/Protobuf/Internal/DescriptorPool.php on line 125

Notice: Undefined index: .simpletest.TestRequest.TestEnum in /Users/trent/src/github.com/sajari/proto-php-example/vendor/google/protobuf/php/src/Google/Protobuf/Internal/DescriptorPool.php on line 125
PHP Notice:  Undefined index:  in /Users/trent/src/github.com/sajari/proto-php-example/vendor/google/protobuf/php/src/Google/Protobuf/Internal/DescriptorPool.php on line 126

Notice: Undefined index:  in /Users/trent/src/github.com/sajari/proto-php-example/vendor/google/protobuf/php/src/Google/Protobuf/Internal/DescriptorPool.php on line 126
Hello Foo!
```

**Comment by mcos** [comment]
@TeBoring: This is still happening. It's due to the fact that `enum` types nested inside messages aren't being registered correctly in the descriptor pool.

```proto
message TestRequest {
  enum TestEnum {
    ZERO = 0;
    ONE = 1;
  }

  string name = 1;

  TestEnum test_enum = 2;
}

```

The `TestEnum` `enum` is being generated as `\SimpleTest\TestRequest_TestEnum`, yet there's no record of it in anything internal to `DescriptorPool`.  This leads me to believe that there's either a bug or an unhandled valid case in [descriptor.php](https://github.com/google/protobuf/blob/d9e0119f2c2f49ea89012a03c030e2b5011d0bc3/php/src/Google/Protobuf/descriptor.php). This is almost certainly a bug in the protobuf syntax parser, and I'd really appreciate if you could give it some priority, as it's not treating valid protobuf syntax correctly.

**Comment by mcos** [comment]
Also, following on from this, there's no test for this situation in https://github.com/google/protobuf/blob/d9e0119f2c2f49ea89012a03c030e2b5011d0bc3/php/tests/proto/test.proto, which I would expect to fail here.

**Comment by mcos** [comment]
@TeBoring any progress on this?

**Comment by TeBoring** [comment]
Sorry, I don't have time to work on this yet. But I do want to include the fix in the 3.3 release. @mcos    Could you send a fix?

**Comment by svdev** [comment]
I'm having the same issue. It seems that nested enums inside messages are not being parsed correctly. I believe this is happening at \Google\Protobuf\Internal\FileDescriptor::buildFromProto($proto). It seems that only the messages/enums in the top level are being parsed. Shouldn't this be a recursive function to be able to parse types for more than 1 level based on the .proto file ?

**Comment by TeBoring** [comment]
Added test in #2989. However, I don't find the bug. @mcos, @svdev, could you verify?
BTW, grpc may depend on old version of protobuf. The latest changes for php is in 3.3.x branch.

**Comment by TeBoring** [comment]
Could you paste your proto?

**Comment by TeBoring** [comment]
test_no_namespace.proto should have covered your case.

**Comment by FrechNina** [comment]
I'm having the same issue and this post is a little bit confused for me:  it is a bug or not?

I have downloaded protobuf 3.2 via composer and run the php test: generated_class_test.php, everything is fine.

I then changed the file test_no_namespace.proto and generated the php files via the same command

syntax = "proto3";

message NoNameSpaceMessage {
  int32 a = 1;

  enum NestedEnum {
    ZERO = 0;
  }
  NestedEnum b = 2;
}

enum NoNameSpaceEnum {
  VALUE_A = 0;
  VALUE_B = 1;
} 

I try to run the generated_class_test.php, but firstly I got

PHP Parse error:  syntax error, unexpected '\' (T_NS_SEPARATOR), expecting identifier (T_STRING) in /var/www/html/devel/src/vendor/google/protobuf/php/tests/generated/NoNameSpaceMessage.php on line 58

I changed the following code
![image](https://cloud.githubusercontent.com/assets/1625807/25268892/bb3f4fcc-267a-11e7-8508-ac29cecd2ae8.png)
to
![image](https://cloud.githubusercontent.com/assets/1625807/25268978/12790210-267b-11e7-99f0-280e845243a2.png)

and try to run the test file: generated_class_test.php

Now, I got:
![image](https://cloud.githubusercontent.com/assets/1625807/25269012/367103f2-267b-11e7-9c6e-01e4c813d72c.png)

**Comment by FrechNina** [comment]
I tried with protobuf 3.3.0 from the 3.3.x branch. I got errors when I try to run the test generated_class_test.php 

![image](https://cloud.githubusercontent.com/assets/1625807/25269517/4e7e4980-267d-11e7-8c06-6ddf9d7c37bc.png)

All errors are the same: Undefined index: .NoNamespaceMessage.NestedEnum

@TeBoring Could you please verify? And I'd really appreciate if you could give it some priority.

**Comment by TeBoring** [comment]
@FrechNina I don't see the error you found. All our test are passed before being merged.

**Comment by TeBoring** [comment]
@FrechNina Could you tell me how to do the test? From where to download the code and how to trigger it.

**Comment by TeBoring** [comment]
Before doing test, you need to:
1) compile protoc
2) use protoc to generate code.

**Comment by TeBoring** [comment]
Feel free to reopen if there is still bug.

**Comment by FrechNina** [comment]
@TeBoring I have completely reinstalled the library. Unfortunately, I got the same errors.

The commands I used:

git clone https://github.com/google/protobuf.git
cd protobuf
git checkout 3.3.x

sudo apt-get install autoconf automake libtool curl make g++ unzip
./autogen.sh
./configure --prefix=/usr
make
make check  # All 7 tests passed
sudo make install
sudo ldconfig

protoc --version #libprotoc 3.3.0
which protoc #/usr/bin/protoc
cd php
composer install
cd tests/
mkdir generated
protoc --php_out=generated proto/*.proto
cd ..
vendor/bin/phpunit tests/generated_class_test.php

**Comment by TeBoring** [comment]
Should have been fixed in #3009
